src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -255,40 +261,149 @@ contract TokenStakingNode is
-            address(this)
-        );
+    function undelegate()
+        public
+        override
+        onlyDelegator
+        onlyWhenSynchronized
+        returns (bytes32[] memory withdrawalRoots)
+    {
+        IDelegationManagerExtended delegationManager =
+            IDelegationManagerExtended(address(tokenStakingNodesManager.delegationManager()));
+
+        (IStrategy[] memory strategies, uint256[] memory shares) = delegationManager.getDelegatableShares(address(this));
+
+        withdrawalRoots = delegationManager.undelegate(address(this));
+
+        // Update queued shares for each strategy
+        for (uint256 i = 0; i < strategies.length; i++) {
+            queuedShares[strategies[i]] += shares[i];
+        }
+
+        delegatedTo = address(0);
 
         emit Undelegated(withdrawalRoots);
     }
 
+    /**
+     * @notice Sets the claimer for rewards using the rewards coordinator
+     * @dev Only callable by delegator. Sets the claimer address for this staking node's rewards.
+     * @param claimer The address to set as the claimer
+     */
+    function setClaimer(address claimer) external onlyDelegator {
+        IRewardsCoordinator rewardsCoordinator = tokenStakingNodesManager.rewardsCoordinator();
+        rewardsCoordinator.setClaimerFor(claimer);
+        emit ClaimerSet(claimer);
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  SYNCHRONIZATION  ---------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice Checks if the StakingNode's delegation state is synced with the DelegationManager.
+     * @dev Compares the locally stored delegatedTo address with the actual delegation in DelegationManager.
+     * @return True if the delegation state is synced, false otherwise.
+     */
+    function isSynchronized() public view returns (bool) {
+        IDelegationManager delegationManager = IDelegationManager(address(tokenStakingNodesManager.delegationManager()));
+        return delegatedTo == delegationManager.delegatedTo(address(this));
+    }
+
+    /**
+     * @notice Synchronizes the staking node's delegation state with the DelegationManager.
+     * @dev This function will be called by the trusted entity when the operator calls undelegate() on this staking node because shares accounting will be out of sync
+     * @param queuedSharesAmounts The amount of shares to be queued for each strategy.
+     * @param undelegateBlockNumber The block number of the undelegate() call by the operator
+     * @param strategies The strategies to be queued.
+     */
+    function synchronize(
+        uint256[] calldata queuedSharesAmounts,
+        uint32 undelegateBlockNumber,
+        IStrategy[] calldata strategies
+    ) public onlyDelegator {
+        if (isSynchronized()) {
+            revert AlreadySynchronized();
+        }
+
```

Comment:
how about if also require delegatedTo == address(0)?

Seems like there should be no other situation where you'd want to call synchronize